hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ReplicationManager.java

```diff
@@ -512,25 +523,60 @@ private void handleUnderReplicatedContainer(final ContainerInfo container,
       if (source.size() > 0) {
         final int replicationFactor = container
             .getReplicationFactor().getNumber();
-        final int delta = replicationFactor - getReplicaCount(id, replicas);
+        // Want to check if the container is mis-replicated after considering
+        // inflight add and delete.
+        // Create a new list from source (healthy replicas minus pending delete)
+        List<DatanodeDetails> targetReplicas = new ArrayList<>(source);
+        // Then add any pending additions
+        targetReplicas.addAll(replicationInFlight);
+
+        int delta = replicationFactor - getReplicaCount(id, replicas);
+        final ContainerPlacementStatus placementStatus =
+            containerPlacement.validateContainerPlacement(
+                targetReplicas, replicationFactor);
+        final int misRepDelta = placementStatus.additionalReplicaRequired();
+        final int replicasNeeded
+            = delta < misRepDelta ? misRepDelta : delta;
+
         final List<DatanodeDetails> excludeList = replicas.stream()
             .map(ContainerReplica::getDatanodeDetails)
             .collect(Collectors.toList());
+        excludeList.addAll(replicationInFlight);
         List<InflightAction> actionList = inflightReplication.get(id);
         if (actionList != null) {
           actionList.stream().map(r -> r.datanode)
               .forEach(excludeList::add);
         }
         final List<DatanodeDetails> selectedDatanodes = containerPlacement
-            .chooseDatanodes(excludeList, null, delta,
+            .chooseDatanodes(excludeList, null, replicasNeeded,
                 container.getUsedBytes());
-
-        LOG.info("Container {} is under replicated. Expected replica count" +
-                " is {}, but found {}.", id, replicationFactor,
-            replicationFactor - delta);
-
-        for (DatanodeDetails datanode : selectedDatanodes) {
-          sendReplicateCommand(container, datanode, source);
+        if (delta > 0) {
+          LOG.info("Container {} is under replicated. Expected replica count" +
+                  " is {}, but found {}.", id, replicationFactor,
+              replicationFactor - delta);
+        }
+        int newMisRepDelta = misRepDelta;
+        if (misRepDelta > 0) {
+          LOG.info("Container: {}. {}",
+              id, placementStatus.misReplicatedReason());
+          // Check if the new target nodes (original plus newly selected nodes)
+          // makes the placement policy valid.
+          targetReplicas.addAll(selectedDatanodes);
+          newMisRepDelta = containerPlacement.validateContainerPlacement(
+              targetReplicas, replicationFactor).additionalReplicaRequired();
```

Comment:
The rack-aware policy's chooseDatanodes() method we use ensures that based on the excluded nodes list we get back a proper list, I believe we do not need to check it twice Are you aware of any cases where this is not true? Based on my understanding of the logic there I would expect all future placement policies to do so, as other existing ones implicitly do so now. If necessary we can add this to the contract of the interface, though this is an interesting property of the rack-aware policy that would worth some re-thinking but that is out of scope for now.

Reply:
> The rack-aware policy's chooseDatanodes() method we use ensures that based on the excluded nodes list we get back a proper list

This is where it gets tricky. There is a fallback (certainly in the pipeline placement policy, but I have not checked the ContainerPlacementRackAware in complete detail) where if the placement policy cannot select enough racks, it will give back nodes which do not meet the placement.

This is to ensure (in the case of pipelinePlacementPolicy) that the cluster can keep on operating if only 1 rack is available for some reason. For the closed container placement, this should be the case too, as if you are down to 1 replica, it is better it gets replicated onto more nodes on the same rack, rather than risk losing the data.

We probably need a future change to scrub the bad pipelines if this fall back happens, while replication manager should deal with closed containers.

Reply:
Right, I see your point... I am not sure, haven't checked, but I guess based on NetworkTopology, and maybe NodeManager we can tell how many healthy racks we have, and if we have one, and the fallback is enabled, then we can report the number of required replicas to be zero in the case we got down to 1 live rack only, while if fallback is not enabled, then we should report 1 additional replica to be required, and give a warning in the logs as the policy can not be met.
I think it would be more clear to handle that in the validation of rack aware policies if we can. What do you think?

Note that we are hardcoding fallback = true at the moment at the only place where we call the policy factory in non-test code.

Reply:
The nodemanager + topology can easily tell us the total racks every registered since SCM was started. But if you have a case where the cluster had 2 racks and than a rack has stopped, that is not so easy to see. You would need to filter all the non-healthy nodes, which is basically what the placement policy does. Then it falls back to give you a non-placement valid node if it needs to. It is important it does this, as it needs to handle under-replication too via the same chooseNodes methods.

The whole area around the fallback needs further thought in general. If we fall to 1 rack, on a cluster that had 2 racks, then its safe to assume the other rack should come back, and until it does the containers are mis-replicated (and probably under-replicated too) and should be reported as such.

In both cases a warning would be written to the logs stating:

```
LOG.warn("Container {} is mis-replicated, requiring {} additional " +
              "replicas. After selecting new nodes, mis-replication has not " +
              "improved. No additional replicas will be scheduled",
              id, misRepDelta);
```
This is really an edge case (2 rack cluster going into 1 rack) and I think we should leave this as it is for now as there are surely other things need to be thought about here in the wider network topology logic.

Reply:
Ok, let's do so for now, but if we can, try to keep this in mind as a possible improvement, as the current check feels a bit hacky still :)

Reply:
Yea, there are improvements needed in this area for sure. I'd like to see how it works in practice before we jump in with too many changes, but its certain on my radar for improvements.